libraries/AP_HAL_Linux/UARTDevice.cpp

```diff
@@ -86,55 +87,89 @@ void UARTDevice::set_blocking(bool blocking)
 
 void UARTDevice::_disable_crlf()
 {
-    struct termios t;
-    memset(&t, 0, sizeof(t));
+    struct termios2 t = { 0 };
 
-    tcgetattr(_fd, &t);
+    if (ioctl(_fd, TCGETS2, &t) != 0) {
+        ::fprintf(stderr, "Failed to read serial options for %s - %s\n",
+                  _device_path, strerror(errno));
+        return;
+    }
 
     // disable LF -> CR/LF
     t.c_iflag &= ~(BRKINT | ICRNL | IMAXBEL | IXON | IXOFF);
     t.c_oflag &= ~(OPOST | ONLCR);
     t.c_lflag &= ~(ISIG | ICANON | IEXTEN | ECHO | ECHOE | ECHOK | ECHOCTL | ECHOKE);
     t.c_cc[VMIN] = 0;
 
-    tcsetattr(_fd, TCSANOW, &t);
+    if (ioctl(_fd, TCSETS2, &t) != 0) {
+        ::fprintf(stderr, "Failed to disable crlf on %s - %s\n",
+                  _device_path, strerror(errno));
+        return;
+    }
 }
 
 void UARTDevice::set_speed(uint32_t baudrate)
 {
-    struct termios t;
-    memset(&t, 0, sizeof(t));
+    struct termios2 tio = { 0 };
 
-    tcgetattr(_fd, &t);
-    cfsetspeed(&t, baudrate);
-    tcsetattr(_fd, TCSANOW, &t);
+    if (ioctl(_fd, TCGETS2, &tio) != 0) {
+        ::fprintf(stderr, "Failed to read serial options for %s - %s\n",
+                  _device_path, strerror(errno));
+        return;
+    }
+
+    // use CBAUDEX to gain access to "non-standard" rates that are common for eg. RC receivers
+    tio.c_cflag &= ~CBAUD;
+    tio.c_cflag |= CBAUDEX;
+    tio.c_ispeed = baudrate;
+    tio.c_ospeed = baudrate;
+
+    if (ioctl(_fd, TCSETS2, &tio) != 0) {
+        ::fprintf(stderr, "Failed to set serial baud to %d for %s - %s\n",
+                  baudrate, _device_path, strerror(errno));
+        return;
+    }
 }
 
 void UARTDevice::set_flow_control(AP_HAL::UARTDriver::flow_control flow_control_setting)
 {
-    struct termios t;
+    struct termios2 t = { 0 };
 
-    if (_flow_control == flow_control_setting) {
+    if (ioctl(_fd, TCGETS2, &t) != 0) {
+        ::fprintf(stderr, "Failed to read serial options for %s - %s\n",
+                  _device_path, strerror(errno));
         return;
     }
 
-    tcgetattr(_fd, &t);
+    if (_flow_control == flow_control_setting) {
```

Comment:
I think this check should be first

Reply:
Ahh yes, good catch. Updated. I also moved the declaration of the struct below that check, which wasn't the way it was before. Not sure if you prefer old C up-top variable declarations. If you do, I'll put it back where it was :smile: